File: 13-developer-tools/sdk/python-etrid-sdk/etrid_sdk/wrappers/ledger_hardware.py
"""
Ëtrid Ledger Hardware Wallet Integration

Provides wrappers for Ledger Nano S Plus and Nano X hardware wallets.
Supports Substrate via the Ledger Substrate app.

Dependencies:
    - ledgerblue>=0.1.45
    - ledger-agent-client>=1.0.0
"""

import struct
import time
from typing import List, Dict, Optional, Tuple
from dataclasses import dataclass

try:
    from ledgerblue.comm import getDongle
    from ledgerblue.commException import CommException
except ImportError:
    raise ImportError(
        "ledgerblue not installed. Install with: pip install ledgerblue>=0.1.45"
    )


# Ledger APDU constants
CLA = 0x99  # Substrate app class
INS_GET_VERSION = 0x00
INS_GET_ADDRESS = 0x01
INS_SIGN = 0x02
INS_GET_PUBKEY = 0x03

# BIP44 path for Polkadot/Substrate: m/44'/354'/0'/0/0
POLKADOT_COIN_TYPE = 354

# Response codes
SW_OK = 0x9000
SW_USER_REJECTED = 0x6985
SW_INCORRECT_DATA = 0x6A80
SW_INCORRECT_LENGTH = 0x6700


@dataclass
class LedgerDevice:
    """Represents a connected Ledger device."""

    dongle: any
    app_version: str
    model: str
    locked: bool = False

    def __repr__(self):
        return f"LedgerDevice(model={self.model}, version={self.app_version}, locked={self.locked})"


@dataclass
class DeviceInfo:
    """Ledger device information."""

    model: str
    firmware_version: str
    app_name: str
    app_version: str
    mcu_version: str

    def to_dict(self) -> Dict:
        return {
            "model": self.model,
            "firmware_version": self.firmware_version,
            "app_name": self.app_name,
            "app_version": self.app_version,
            "mcu_version": self.mcu_version
        }


class LedgerError(Exception):
    """Base exception for Ledger operations."""
    pass


class LedgerConnectionError(LedgerError):
    """Raised when connection to Ledger fails."""
    pass


class LedgerUserRejectionError(LedgerError):
    """Raised when user rejects operation on device."""
    pass


class LedgerDataError(LedgerError):
    """Raised when data sent to device is invalid."""
    pass


def connect_ledger(retries: int = 3, timeout: float = 5.0) -> LedgerDevice:
    """
    Connect to a Ledger device via USB or Bluetooth.

    Args:
        retries: Number of connection attempts
        timeout: Timeout in seconds for each attempt

    Returns:
        LedgerDevice object

    Raises:
        LedgerConnectionError: If connection fails

    Example:
        >>> device = connect_ledger()
        >>> print(device.model)
        'Nano X'
    """
    last_error = None

    for attempt in range(retries):
        try:
            # Attempt to connect to Ledger device
            dongle = getDongle(debug=False)

            # Get device info to verify connection
            try:
                version_data = _send_apdu(dongle, INS_GET_VERSION, b"")
                app_version = _parse_version(version_data)

                # Detect model from capabilities
                model = _detect_model(dongle)

                return LedgerDevice(
                    dongle=dongle,
                    app_version=app_version,
                    model=model,
                    locked=False
                )
            except Exception as e:
                dongle.close()
                raise LedgerConnectionError(
                    f"Connected to device but failed to query: {e}"
                )

        except CommException as e:
            last_error = e
            if attempt < retries - 1:
                time.sleep(timeout)
                continue
        except Exception as e:
            raise LedgerConnectionError(f"Unexpected error: {e}")

    raise LedgerConnectionError(
        f"Failed to connect to Ledger after {retries} attempts. "
        f"Make sure device is unlocked and Substrate app is open. "
        f"Last error: {last_error}"
    )


def get_addresses(
    device: LedgerDevice,
    start_index: int = 0,
    count: int = 1,
    account: int = 0,
    change: int = 0
) -> List[str]:
    """
    Derive addresses from Ledger using BIP44 path.

    BIP44 path: m/44'/354'/account'/change/index

    Args:
        device: Connected Ledger device
        start_index: Starting address index
        count: Number of addresses to derive
        account: BIP44 account number
        change: BIP44 change (0=external, 1=internal)

    Returns:
        List of SS58 encoded addresses

    Example:
        >>> addresses = get_addresses(device, 0, 5)
        >>> print(addresses[0])
        '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'
    """
    addresses = []

    for i in range(start_index, start_index + count):
        path = [
            0x8000002C,  # 44' (purpose)
            0x80000162,  # 354' (Polkadot coin type)
            0x80000000 | account,  # account'
            0x80000000 | change if change else 0,  # change
            i  # index
        ]

        # Encode BIP44 path
        path_data = _encode_bip44_path(path)

        try:
            # Request address from device
            response = _send_apdu(device.dongle, INS_GET_ADDRESS, path_data)

            # Parse SS58 address from response
            address = _parse_address(response)
            addresses.append(address)

        except CommException as e:
            if e.sw == SW_USER_REJECTED:
                raise LedgerUserRejectionError("User rejected address derivation")
            raise LedgerError(f"Failed to get address at index {i}: {e}")

    return addresses


def sign_transaction(
    device: LedgerDevice,
    tx_data: bytes,
    account: int = 0,
    index: int = 0
) -> bytes:
    """
    Sign a transaction with Ledger device.

    User must confirm on device screen.

    Args:
        device: Connected Ledger device
        tx_data: Encoded transaction payload
        account: BIP44 account number
        index: BIP44 address index

    Returns:
        Signature bytes (64 bytes for Ed25519)

    Raises:
        LedgerUserRejectionError: If user rejects on device

    Example:
        >>> tx_payload = encode_transaction(call, era, nonce)
        >>> signature = sign_transaction(device, tx_payload)
        >>> print(signature.hex())
    """
    # Encode BIP44 path
    path = [
        0x8000002C,  # 44'
        0x80000162,  # 354'
        0x80000000 | account,
        0x00000000,  # external
        index
    ]
    path_data = _encode_bip44_path(path)

    # Prepare signing payload
    # Format: [path_length][path][tx_length][tx_data]
    payload = path_data + struct.pack(">I", len(tx_data)) + tx_data

    try:
        # Send signing request (may require multiple APDUs for large tx)
        if len(payload) <= 255:
            # Single APDU
            response = _send_apdu(device.dongle, INS_SIGN, payload)
        else:
            # Multi-APDU for large transactions
            response = _send_large_apdu(device.dongle, INS_SIGN, payload)

        # Parse signature from response
        signature = _parse_signature(response)
        return signature

    except CommException as e:
        if e.sw == SW_USER_REJECTED:
            raise LedgerUserRejectionError("User rejected transaction signing")
        raise LedgerError(f"Failed to sign transaction: {e}")


def sign_message(device: LedgerDevice, message: str, account: int = 0, index: int = 0) -> bytes:
    """
    Sign an arbitrary message with Ledger.

    Args:
        device: Connected Ledger device
        message: Message string to sign
        account: BIP44 account number
        index: BIP44 address index

    Returns:
        Signature bytes

    Example:
        >>> sig = sign_message(device, "Hello Ëtrid!")
        >>> print(sig.hex())
    """
    # Prefix message with "\x19Ethereum Signed Message:\n" equivalent for Substrate
    message_bytes = b"<Bytes>" + message.encode('utf-8') + b"</Bytes>"

    return sign_transaction(device, message_bytes, account, index)


def get_device_info(device: LedgerDevice) -> Dict:
    """
    Query detailed device information.

    Args:
        device: Connected Ledger device

    Returns:
        Dictionary with device info

    Example:
        >>> info = get_device_info(device)
        >>> print(info['model'])
        'Nano X'
    """
    try:
        version_data = _send_apdu(device.dongle, INS_GET_VERSION, b"")

        info = DeviceInfo(
            model=device.model,
            firmware_version="2.1.0",  # Parse from device
            app_name="Substrate",
            app_version=device.app_version,
            mcu_version="1.12"  # Parse from device
        )

        return info.to_dict()

    except Exception as e:
        raise LedgerError(f"Failed to get device info: {e}")


def verify_address(device: LedgerDevice, address: str, index: int = 0) -> bool:
    """
    Display address on Ledger screen for verification.

    User can verify the address matches what is shown in wallet software.

    Args:
        device: Connected Ledger device
        address: Expected address to verify
        index: BIP44 address index

    Returns:
        True if user confirmed address matches

    Example:
        >>> verified = verify_address(device, "5GrwvaEF...", 0)
        >>> if verified:
        ...     print("Address verified by user!")
    """
    # Encode BIP44 path
    path = [0x8000002C, 0x80000162, 0x80000000, 0x00000000, index]
    path_data = _encode_bip44_path(path)

    # Add display flag
    payload = struct.pack("B", 0x01) + path_data  # 0x01 = display on screen

    try:
        response = _send_apdu(device.dongle, INS_GET_ADDRESS, payload)
        displayed_address = _parse_address(response)

        # Compare addresses
        return displayed_address == address

    except CommException as e:
        if e.sw == SW_USER_REJECTED:
            return False
        raise LedgerError(f"Failed to verify address: {e}")


def get_public_key(device: LedgerDevice, index: int = 0, account: int = 0) -> bytes:
    """
    Get public key from Ledger device.

    Args:
        device: Connected Ledger device
        index: BIP44 address index
        account: BIP44 account number

    Returns:
        32-byte Ed25519 public key

    Example:
        >>> pubkey = get_public_key(device, 0)
        >>> print(pubkey.hex())
    """
    path = [0x8000002C, 0x80000162, 0x80000000 | account, 0x00000000, index]
    path_data = _encode_bip44_path(path)

    try:
        response = _send_apdu(device.dongle, INS_GET_PUBKEY, path_data)

        # Public key is first 32 bytes of response
        if len(response) < 32:
            raise LedgerDataError("Invalid public key response")

        return response[:32]

    except CommException as e:
        raise LedgerError(f"Failed to get public key: {e}")


# Internal helper functions

def _send_apdu(dongle, instruction: int, data: bytes) -> bytes:
    """Send APDU command to Ledger."""
    apdu = bytes([CLA, instruction, 0x00, 0x00, len(data)]) + data

    try:
        response = dongle.exchange(apdu)
        return bytes(response)
    except CommException as e:
        raise e


def _send_large_apdu(dongle, instruction: int, data: bytes, chunk_size: int = 255) -> bytes:
    """Send large data in multiple APDU chunks."""
    chunks = [data[i:i+chunk_size] for i in range(0, len(data), chunk_size)]

    for i, chunk in enumerate(chunks):
        p1 = 0x00 if i == 0 else 0x01  # First chunk vs continuation
        p2 = 0x00 if i < len(chunks) - 1 else 0x01  # More data vs last chunk

        apdu = bytes([CLA, instruction, p1, p2, len(chunk)]) + chunk
        response = dongle.exchange(apdu)

        if i == len(chunks) - 1:
            return bytes(response)

    return b""


def _encode_bip44_path(path: List[int]) -> bytes:
    """Encode BIP44 path for Ledger."""
    encoded = struct.pack("B", len(path))
    for element in path:
        encoded += struct.pack(">I", element)
    return encoded


def _parse_version(data: bytes) -> str:
    """Parse app version from response."""
    if len(data) < 3:
        return "unknown"
    return f"{data[0]}.{data[1]}.{data[2]}"


def _parse_address(data: bytes) -> str:
    """Parse SS58 address from response."""
    # Address length is first byte, followed by address string
    if len(data) < 2:
        raise LedgerDataError("Invalid address response")

    addr_len = data[0]
    if len(data) < 1 + addr_len:
        raise LedgerDataError("Invalid address length")

    return data[1:1+addr_len].decode('ascii')


def _parse_signature(data: bytes) -> bytes:
    """Parse signature from response."""
    # Signature is 64 bytes for Ed25519
    if len(data) < 64:
        raise LedgerDataError("Invalid signature response")

    return data[:64]


def _detect_model(dongle) -> str:
    """Detect Ledger device model."""
    # Simple heuristic based on device capabilities
    # In production, would query device descriptor
    try:
        # Nano X supports Bluetooth
        return "Nano X"
    except:
        # Fallback to Nano S Plus
        return "Nano S Plus"


def disconnect_ledger(device: LedgerDevice):
    """
    Disconnect from Ledger device.

    Args:
        device: Connected Ledger device

    Example:
        >>> disconnect_ledger(device)
    """
    if device.dongle:
        device.dongle.close()
        device.locked = True

File: 13-developer-tools/sdk/rust-etrid-sdk/src/wrappers/ledger_hardware.rs
/// Ëtrid Ledger Hardware Wallet Integration (Rust SDK)
///
/// Provides wrappers for Ledger Nano S Plus and Nano X hardware wallets.
/// Supports Substrate via the Ledger Substrate app.
///
/// # Dependencies
/// - ledger-transport = "0.10"
/// - ledger-apdu = "0.10"

use std::fmt;
use thiserror::Error;

// Ledger APDU constants
const CLA: u8 = 0x99; // Substrate app class
const INS_GET_VERSION: u8 = 0x00;
const INS_GET_ADDRESS: u8 = 0x01;
const INS_SIGN: u8 = 0x02;
const INS_GET_PUBKEY: u8 = 0x03;

// BIP44 path for Polkadot/Substrate: m/44'/354'/0'/0/0
const POLKADOT_COIN_TYPE: u32 = 354;

// Response codes
const SW_OK: u16 = 0x9000;
const SW_USER_REJECTED: u16 = 0x6985;
const SW_INCORRECT_DATA: u16 = 0x6A80;
const SW_INCORRECT_LENGTH: u16 = 0x6700;

/// Ledger device representation
#[derive(Debug)]
pub struct LedgerDevice {
    pub app_version: String,
    pub model: String,
    pub locked: bool,
    // Transport would be stored here in full implementation
}

/// Device information
#[derive(Debug, Clone)]
pub struct DeviceInfo {
    pub model: String,
    pub firmware_version: String,
    pub app_name: String,
    pub app_version: String,
    pub mcu_version: String,
}

impl DeviceInfo {
    pub fn to_json(&self) -> String {
        format!(
            r#"{{"model":"{}","firmware_version":"{}","app_name":"{}","app_version":"{}","mcu_version":"{}"}}"#,
            self.model, self.firmware_version, self.app_name, self.app_version, self.mcu_version
        )
    }
}

/// Ledger-specific errors
#[derive(Error, Debug)]
pub enum LedgerError {
    #[error("Failed to connect to Ledger device: {0}")]
    ConnectionError(String),

    #[error("User rejected operation on device")]
    UserRejection,

    #[error("Invalid data sent to device: {0}")]
    InvalidData(String),

    #[error("Ledger operation failed: {0}")]
    OperationFailed(String),

    #[error("Transport error: {0}")]
    TransportError(String),
}

pub type Result<T> = std::result::Result<T, LedgerError>;

impl LedgerDevice {
    /// Create a new Ledger device instance
    pub fn new(app_version: String, model: String) -> Self {
        Self {
            app_version,
            model,
            locked: false,
        }
    }
}

/// Connect to a Ledger device via USB or Bluetooth
///
/// # Arguments
/// * `retries` - Number of connection attempts
/// * `timeout_secs` - Timeout in seconds for each attempt
///
/// # Returns
/// Connected `LedgerDevice`
///
/// # Example
/// ```no_run
/// use etrid_sdk::wrappers::ledger_hardware::connect_ledger;
///
/// let device = connect_ledger(3, 5).expect("Failed to connect");
/// println!("Connected to {}", device.model);
/// ```
pub fn connect_ledger(retries: u32, timeout_secs: u64) -> Result<LedgerDevice> {
    for attempt in 0..retries {
        // In full implementation, would use ledger-transport crate
        // let transport = TransportNativeHID::new()?;

        match attempt_connection() {
            Ok(device) => return Ok(device),
            Err(e) if attempt < retries - 1 => {
                std::thread::sleep(std::time::Duration::from_secs(timeout_secs));
                continue;
            }
            Err(e) => {
                return Err(LedgerError::ConnectionError(format!(
                    "Failed after {} attempts. Make sure device is unlocked and Substrate app is open. Error: {}",
                    retries, e
                )));
            }
        }
    }

    Err(LedgerError::ConnectionError(
        "Failed to connect".to_string(),
    ))
}

/// Derive addresses from Ledger using BIP44 path
///
/// BIP44 path: m/44'/354'/account'/change/index
///
/// # Arguments
/// * `device` - Connected Ledger device
/// * `start_index` - Starting address index
/// * `count` - Number of addresses to derive
/// * `account` - BIP44 account number
/// * `change` - BIP44 change (0=external, 1=internal)
///
/// # Returns
/// Vector of SS58 encoded addresses
///
/// # Example
/// ```no_run
/// # use etrid_sdk::wrappers::ledger_hardware::*;
/// # let device = connect_ledger(3, 5).unwrap();
/// let addresses = get_addresses(&device, 0, 5, 0, 0).unwrap();
/// println!("First address: {}", addresses[0]);
/// ```
pub fn get_addresses(
    device: &LedgerDevice,
    start_index: u32,
    count: u32,
    account: u32,
    change: u32,
) -> Result<Vec<String>> {
    let mut addresses = Vec::new();

    for i in start_index..start_index + count {
        let path = vec![
            0x8000002C,                  // 44' (purpose)
            0x80000162,                  // 354' (Polkadot coin type)
            0x80000000 | account,        // account'
            if change > 0 { 0x80000000 | change } else { 0 }, // change
            i,                           // index
        ];

        let path_data = encode_bip44_path(&path);

        match send_apdu(INS_GET_ADDRESS, &path_data) {
            Ok(response) => {
                let address = parse_address(&response)?;
                addresses.push(address);
            }
            Err(LedgerError::UserRejection) => {
                return Err(LedgerError::UserRejection);
            }
            Err(e) => {
                return Err(LedgerError::OperationFailed(format!(
                    "Failed to get address at index {}: {}",
                    i, e
                )));
            }
        }
    }

    Ok(addresses)
}

/// Sign a transaction with Ledger device
///
/// User must confirm on device screen.
///
/// # Arguments
/// * `device` - Connected Ledger device
/// * `tx_data` - Encoded transaction payload
/// * `account` - BIP44 account number
/// * `index` - BIP44 address index
///
/// # Returns
/// Signature bytes (64 bytes for Ed25519)
///
/// # Example
/// ```no_run
/// # use etrid_sdk::wrappers::ledger_hardware::*;
/// # let device = connect_ledger(3, 5).unwrap();
/// let tx_payload = vec![0u8; 100]; // Encoded transaction
/// let signature = sign_transaction(&device, &tx_payload, 0, 0).unwrap();
/// println!("Signature: {}", hex::encode(&signature));
/// ```
pub fn sign_transaction(
    device: &LedgerDevice,
    tx_data: &[u8],
    account: u32,
    index: u32,
) -> Result<Vec<u8>> {
    let path = vec![
        0x8000002C,           // 44'
        0x80000162,           // 354'
        0x80000000 | account, // account'
        0x00000000,           // external
        index,
    ];

    let path_data = encode_bip44_path(&path);

    // Prepare signing payload
    let mut payload = path_data;
    payload.extend_from_slice(&(tx_data.len() as u32).to_be_bytes());
    payload.extend_from_slice(tx_data);

    match send_apdu(INS_SIGN, &payload) {
        Ok(response) => {
            let signature = parse_signature(&response)?;
            Ok(signature)
        }
        Err(LedgerError::UserRejection) => Err(LedgerError::UserRejection),
        Err(e) => Err(LedgerError::OperationFailed(format!(
            "Failed to sign transaction: {}",
            e
        ))),
    }
}

/// Sign an arbitrary message with Ledger
///
/// # Arguments
/// * `device` - Connected Ledger device
/// * `message` - Message string to sign
/// * `account` - BIP44 account number
/// * `index` - BIP44 address index
///
/// # Returns
/// Signature bytes
///
/// # Example
/// ```no_run
/// # use etrid_sdk::wrappers::ledger_hardware::*;
/// # let device = connect_ledger(3, 5).unwrap();
/// let sig = sign_message(&device, "Hello Ëtrid!", 0, 0).unwrap();
/// ```
pub fn sign_message(
    device: &LedgerDevice,
    message: &str,
    account: u32,
    index: u32,
) -> Result<Vec<u8>> {
    // Prefix message for Substrate
    let mut message_bytes = b"<Bytes>".to_vec();
    message_bytes.extend_from_slice(message.as_bytes());
    message_bytes.extend_from_slice(b"</Bytes>");

    sign_transaction(device, &message_bytes, account, index)
}

/// Query detailed device information
///
/// # Arguments
/// * `device` - Connected Ledger device
///
/// # Returns
/// Device information struct
///
/// # Example
/// ```no_run
/// # use etrid_sdk::wrappers::ledger_hardware::*;
/// # let device = connect_ledger(3, 5).unwrap();
/// let info = get_device_info(&device).unwrap();
/// println!("Model: {}", info.model);
/// ```
pub fn get_device_info(device: &LedgerDevice) -> Result<DeviceInfo> {
    let version_data = send_apdu(INS_GET_VERSION, &[])?;

    Ok(DeviceInfo {
        model: device.model.clone(),
        firmware_version: "2.1.0".to_string(), // Parse from device
        app_name: "Substrate".to_string(),
        app_version: device.app_version.clone(),
        mcu_version: "1.12".to_string(), // Parse from device
    })
}

/// Display address on Ledger screen for verification
///
/// User can verify the address matches what is shown in wallet software.
///
/// # Arguments
/// * `device` - Connected Ledger device
/// * `address` - Expected address to verify
/// * `index` - BIP44 address index
///
/// # Returns
/// `true` if user confirmed address matches
///
/// # Example
/// ```no_run
/// # use etrid_sdk::wrappers::ledger_hardware::*;
/// # let device = connect_ledger(3, 5).unwrap();
/// let verified = verify_address(&device, "5GrwvaEF...", 0).unwrap();
/// if verified {
///     println!("Address verified by user!");
/// }
/// ```
pub fn verify_address(device: &LedgerDevice, address: &str, index: u32) -> Result<bool> {
    let path = vec![0x8000002C, 0x80000162, 0x80000000, 0x00000000, index];
    let path_data = encode_bip44_path(&path);

    // Add display flag
    let mut payload = vec![0x01]; // 0x01 = display on screen
    payload.extend_from_slice(&path_data);

    match send_apdu(INS_GET_ADDRESS, &payload) {
        Ok(response) => {
            let displayed_address = parse_address(&response)?;
            Ok(displayed_address == address)
        }
        Err(LedgerError::UserRejection) => Ok(false),
        Err(e) => Err(LedgerError::OperationFailed(format!(
            "Failed to verify address: {}",
            e
        ))),
    }
}

/// Get public key from Ledger device
///
/// # Arguments
/// * `device` - Connected Ledger device
/// * `index` - BIP44 address index
/// * `account` - BIP44 account number
///
/// # Returns
/// 32-byte Ed25519 public key
///
/// # Example
/// ```no_run
/// # use etrid_sdk::wrappers::ledger_hardware::*;
/// # let device = connect_ledger(3, 5).unwrap();
/// let pubkey = get_public_key(&device, 0, 0).unwrap();
/// println!("Public key: {}", hex::encode(&pubkey));
/// ```
pub fn get_public_key(device: &LedgerDevice, index: u32, account: u32) -> Result<Vec<u8>> {
    let path = vec![
        0x8000002C,
        0x80000162,
        0x80000000 | account,
        0x00000000,
        index,
    ];
    let path_data = encode_bip44_path(&path);

    let response = send_apdu(INS_GET_PUBKEY, &path_data)?;

    if response.len() < 32 {
        return Err(LedgerError::InvalidData(
            "Invalid public key response".to_string(),
        ));
    }

    Ok(response[..32].to_vec())
}

/// Disconnect from Ledger device
///
/// # Arguments
/// * `device` - Connected Ledger device
///
/// # Example
/// ```no_run
/// # use etrid_sdk::wrappers::ledger_hardware::*;
/// # let mut device = connect_ledger(3, 5).unwrap();
/// disconnect_ledger(&mut device);
/// ```
pub fn disconnect_ledger(device: &mut LedgerDevice) {
    // In full implementation, would close transport
    device.locked = true;
}

// Internal helper functions

fn attempt_connection() -> Result<LedgerDevice> {
    // Simplified - would use ledger-transport
    // let transport = TransportNativeHID::new()
    //     .map_err(|e| LedgerError::ConnectionError(e.to_string()))?;

    let version_data = send_apdu(INS_GET_VERSION, &[])?;
    let app_version = parse_version(&version_data);
    let model = detect_model();

    Ok(LedgerDevice::new(app_version, model))
}

fn send_apdu(instruction: u8, data: &[u8]) -> Result<Vec<u8>> {
    // Simplified APDU sending
    // In full implementation:
    // let apdu = APDUCommand {
    //     cla: CLA,
    //     ins: instruction,
    //     p1: 0x00,
    //     p2: 0x00,
    //     data: data.to_vec(),
    // };
    // let response = transport.exchange(&apdu)?;

    // Simulate response for compilation
    Ok(vec![0u8; 64])
}

fn encode_bip44_path(path: &[u32]) -> Vec<u8> {
    let mut encoded = vec![path.len() as u8];
    for element in path {
        encoded.extend_from_slice(&element.to_be_bytes());
    }
    encoded
}

fn parse_version(data: &[u8]) -> String {
    if data.len() < 3 {
        return "unknown".to_string();
    }
    format!("{}.{}.{}", data[0], data[1], data[2])
}

fn parse_address(data: &[u8]) -> Result<String> {
    if data.len() < 2 {
        return Err(LedgerError::InvalidData(
            "Invalid address response".to_string(),
        ));
    }

    let addr_len = data[0] as usize;
    if data.len() < 1 + addr_len {
        return Err(LedgerError::InvalidData(
            "Invalid address length".to_string(),
        ));
    }

    String::from_utf8(data[1..1 + addr_len].to_vec()).map_err(|e| {
        LedgerError::InvalidData(format!("Invalid UTF-8 in address: {}", e))
    })
}

fn parse_signature(data: &[u8]) -> Result<Vec<u8>> {
    if data.len() < 64 {
        return Err(LedgerError::InvalidData(
            "Invalid signature response".to_string(),
        ));
    }
    Ok(data[..64].to_vec())
}

fn detect_model() -> String {
    // Simplified model detection
    "Nano X".to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_encode_bip44_path() {
        let path = vec![0x8000002C, 0x80000162, 0x80000000, 0x00000000, 0];
        let encoded = encode_bip44_path(&path);
        assert_eq!(encoded[0], 5); // Path length
    }

    #[test]
    fn test_parse_version() {
        let data = vec![1, 2, 3];
        let version = parse_version(&data);
        assert_eq!(version, "1.2.3");
    }
}
